pallets/capacity/src/types.rs

```diff
@@ -0,0 +1,183 @@
+};
+use scale_info::TypeInfo;
+use sp_runtime::traits::{CheckedAdd, CheckedSub, Saturating, Zero};
+
+#[cfg(any(feature = "runtime-benchmarks", test))]
+use sp_std::vec::Vec;
+
+/// The type used for storing information about staking details.
+#[derive(
+	TypeInfo, RuntimeDebugNoBound, PartialEqNoBound, EqNoBound, Clone, Decode, Encode, MaxEncodedLen,
+)]
+#[scale_info(skip_type_params(T))]
+pub struct StakingAccountDetails<T: Config> {
+	/// The amount a Staker has staked, minus the sum of all tokens in `unlocking`.
+	pub active: BalanceOf<T>,
+	/// The total amount of tokens in `active` and `unlocking`
+	pub total: BalanceOf<T>,
+	/// Unstaked balances that are thawing or awaiting withdrawal.
+	pub unlocking: BoundedVec<UnlockChunk<BalanceOf<T>, T::EpochNumber>, T::MaxUnlockingChunks>,
+}
+/// The type that is used to record a single request for a number of tokens to be unlocked.
+#[derive(PartialEq, Eq, Clone, Encode, Decode, RuntimeDebug, TypeInfo, MaxEncodedLen)]
+pub struct UnlockChunk<Balance, EpochNumber> {
+	/// Amount to be unlocked.
+	pub value: Balance,
+	/// Block number at which point funds are unlocked.
+	pub thaw_at: EpochNumber,
+}
+
+impl<T: Config> StakingAccountDetails<T> {
+	/// Increases total and active balances by an amount.
+	pub fn increase_by(&mut self, amount: BalanceOf<T>) -> Option<()> {
+		self.total = amount.checked_add(&self.total)?;
+		self.active = amount.checked_add(&self.active)?;
+
+		Some(())
+	}
+
+	/// Calculates a stakable amount from a propose amount.
+	pub fn get_stakable_amount_for(
+		&self,
+		staker: &T::AccountId,
+		proposed_amount: BalanceOf<T>,
+	) -> BalanceOf<T> {
+		let account_balance = T::Currency::free_balance(&staker);
+		let available_staking_balance = account_balance.saturating_sub(self.total);
+		available_staking_balance.min(proposed_amount)
+	}
+
+	#[cfg(any(feature = "runtime-benchmarks", test))]
+	///  tmp fn for testing only
+	/// set unlock chunks with (balance, thaw_at).  does not check that the unlock chunks
+	/// don't exceed total.
+	/// returns true on success, false on failure (?)
+	// TODO: remove this when finished and use production fn
+	pub fn set_unlock_chunks(&mut self, chunks: &Vec<(u32, u32)>) -> bool {
+		let result: Vec<UnlockChunk<BalanceOf<T>, <T>::EpochNumber>> = chunks
+			.into_iter()
+			.map(|chunk| UnlockChunk { value: chunk.0.into(), thaw_at: chunk.1.into() })
+			.collect();
+		self.unlocking = BoundedVec::try_from(result).unwrap();
+		self.unlocking.len() == chunks.len()
+	}
+
+	/// deletes thawed chunks, updates `total`, Caller is responsible for updating free/locked
+	/// balance on the token account.
+	/// Returns: the total amount reaped from `unlocking`
+	pub fn reap_thawed(&mut self, current_epoch: <T>::EpochNumber) -> BalanceOf<T> {
+		let mut total_reaped: BalanceOf<T> = 0u32.into();
+		self.unlocking.retain(|chunk| {
+			if current_epoch.ge(&chunk.thaw_at) {
+				total_reaped = total_reaped + chunk.value;
```

Comment:
nit: can this overflow?

Reply:
If not please add comment why

Reply:
good catch, I should have done a checked_add